Add parallelization capability to RandomWpMethodEQOracle

c-southwest opened the pull request:
With this PR, we can specify how many threads we want to parallelize the **find counter example** phase in `RandomWpMethodEQOracle`.

Every XX-Fuzzer wants to use this parallelization need to override `SulConfig.java` interface  `cloneWithThreadId` method, which should return an independent sulConfig object that different threads will use. For example, for DTLS-Fuzzer, different threads should listen to different UDP ports, otherwise will have conflict, and that's why I added this method into the `SulConfig.java` interface.

I also provide a default implementation (just throw exception) for the  `cloneWithThreadId` method, so it should not break compilation for other projects.

Added new parameter:
`-threadCount` or `-tc`

This is not ready for merge yet, I didn't check how statistics works and I will look at it later this week.

I updated DTLS-Fuzzer with this branch, and you can see results in DTLS-Fuzzer PR: https://github.com/assist-project/dtls-fuzzer/pull/190

kostis commented on src/main/java/com/github/protocolfuzzing/protocolstatefuzzer/statefuzzer/core/StateFuzzerComposerStandard.java:
```diff
@@ -116,24 +118,35 @@ public StateFuzzerComposerStandard(
-                .setLoggingWrapper("")
-                .getWrappedSul();
+        SulConfig sulConfig = stateFuzzerEnabler.getSulConfig();
+        for (Integer i = 0; i < sulConfig.getThreadCount(); i++) {
```
`int` instead of `Integer` here?

kostis reviewed (changes requested):
Please see the comments I've left in the code.

kostis commented:
I think this PR looks good overall.

I'll leave it open for some days in case @actyp also wants to provide some review before it gets merged.

actyp reviewed (changes requested):
Hello, I have been summoned!

It is a very interesting addition and hopefully it can save much time spent in the equivalence queries.

Most of my comments are about change minimization and not that critical, but the most important one is about the waiting technique of the `main` thread in `RandomWpMethodEQOracle`. In any instance I may be missing something, so feel free to correct me.

In general the PR is in good shape, just check if my recommendations can enhance it further!

actyp commented on src/main/java/com/github/protocolfuzzing/protocolstatefuzzer/components/learner/oracles/RandomWpMethodEQOracle.java:
```diff
@@ -127,28 +135,61 @@ public RandomWpMethodEQOracle(MealyMembershipOracle<I, O> sulOracle,
+                // process results
+                for (Future<DefaultQuery<I, Word<O>>> future : futures) {
+                    try {
+                        DefaultQuery<I, Word<O>> counterExample = future.get();
```
Thank you for the clarification.

The question then becomes: since we wait for all the threads and we currently discard other possible counterexamples, can we gather a list of unique counterexamples and refine the hypothesis multiple times in `StateFuzzerStandard` instead of a single time ?

What do you think ?

c-southwest commented on src/main/java/com/github/protocolfuzzing/protocolstatefuzzer/statefuzzer/core/StateFuzzerComposerStandard.java:
```diff
@@ -116,30 +121,56 @@ public StateFuzzerComposerStandard(
 
         // initialize the output for the socket closed
-        this.socketClosedOutput = abstractSul.getMapper().getOutputBuilder().buildOutputExact(OutputBuilder.SOCKET_CLOSED);
+        this.socketClosedOutput = sulBuilder.build(sulConfig, cleanupTasks).getMapper().getOutputBuilder().buildOutputExact(OutputBuilder.SOCKET_CLOSED);
```
there is no good reason, just because `abstractSul` is a local variable inside a `for` loop and cannot be used outside.

c-southwest commented on src/main/java/com/github/protocolfuzzing/protocolstatefuzzer/statefuzzer/core/StateFuzzerComposerStandard.java:
```diff
@@ -270,21 +301,24 @@ protected void composeLearner(List<O> terminatingOutputs) {
     protected void composeEquivalenceOracle(List<O> terminatingOutputs) {
+        List<MembershipOracle.MealyMembershipOracle<I, O>> equivalenceSulOracles = new ArrayList<>();
+        for (SUL<I, O> sul : suls) {
+            MembershipOracle.MealyMembershipOracle<I, O> equivalenceSulOracle = new SULOracle<>(sul);
```
Yes, we can use any SUL to run the initialization, refinement, and finding counterexamples.

However, in the PR I only parallelized `finding counterexamples`. I didn't parallelize **initialization** part is because I think it is not very worth, maybe we can save 10 seconds, but we need many parallel related code changes.

For the **refinement** part, I just don't know how it works, I tried to debug it, but the process is strange for me. Refinement process will utilize SUL at many different places, so I have no clue when to parallelize.

c-southwest commented on src/main/java/com/github/protocolfuzzing/protocolstatefuzzer/components/learner/config/LearnerConfigStandard.java:
```diff
@@ -278,6 +279,23 @@ public class LearnerConfigStandard implements LearnerConfig {
+    public SulConfig getSulConfig() {
+        return sulConfig;
+    }
+
```
fixed

actyp replied:
Thanks, just a trivial detail, add a `-eqvThreads` as a second option and correct the typo in the doc sentence `Stores the JCommander Parameter -equivalenceThreadCount, -eqvThreads`

actyp replied:
OK, so as @kostis mentions, @c-southwest let's do the above change now, in order to add a shortened version of the parameter as well.

c-southwest commented on src/main/java/com/github/protocolfuzzing/protocolstatefuzzer/components/learner/config/LearnerConfig.java:
```diff
@@ -13,6 +14,17 @@
+    default SulConfig getSulConfig() {
+        return null;
+    }
+
```
fixed

c-southwest commented on src/main/java/com/github/protocolfuzzing/protocolstatefuzzer/components/learner/factory/LearningSetupFactory.java:
```diff
@@ -215,49 +215,49 @@ public static IOEquivalenceOracle createEquivalenceOracle(
-                        sulOracle, config.getMinLength(), config.getRandLength(),
-                        config.getEquivQueryBound(), config.getSeed());
+                        sulOracles, config.getMinLength(), config.getRandLength(),
+                        config.getEquivQueryBound(), config.getSeed(), config.getSulConfig().getThreadCount());
```
fixed

c-southwest commented on src/main/java/com/github/protocolfuzzing/protocolstatefuzzer/components/sul/core/config/SulConfig.java:
```diff
@@ -12,6 +12,26 @@
+     */
+    default int getThreadCount() {
+        return 1;
+    }
```
fixed

c-southwest commented on src/main/java/com/github/protocolfuzzing/protocolstatefuzzer/statefuzzer/core/StateFuzzerComposerStandard.java:
```diff
@@ -116,30 +121,56 @@ public StateFuzzerComposerStandard(
+        if (this.learnerConfig instanceof LearnerConfigStandard) {
+            ((LearnerConfigStandard) this.learnerConfig).setSulConfig(stateFuzzerEnabler.getSulConfig());
+        }
+
```
fixed

c-southwest commented on src/main/java/com/github/protocolfuzzing/protocolstatefuzzer/statefuzzer/core/StateFuzzerComposerStandard.java:
```diff
@@ -116,30 +121,56 @@ public StateFuzzerComposerStandard(
+                currentSulWrapper = new SulWrapperStandard<>();
+            }
+
+            var sul = currentSulWrapper
```
fixed

c-southwest commented on src/main/java/com/github/protocolfuzzing/protocolstatefuzzer/statefuzzer/core/StateFuzzerStandard.java:
```diff
@@ -131,6 +131,7 @@ protected LearnerResult<MealyMachineWrapper<I, O>> inferStateMachine() {
 
         try {
             LOGGER.info("Input alphabet: {}", alphabet);
+            LOGGER.info("Thread Count: {}", stateFuzzerEnabler.getSulConfig().getThreadCount());
```
fixed

c-southwest commented on src/main/java/com/github/protocolfuzzing/protocolstatefuzzer/components/learner/oracles/RandomWpMethodEQOracle.java:
```diff
@@ -127,28 +135,61 @@ public RandomWpMethodEQOracle(MealyMembershipOracle<I, O> sulOracle,
+                List<Future<DefaultQuery<I, Word<O>>>> futures = new ArrayList<>(threadCount);
+                for (int i = 0; i < threadCount; i++) {
+                    final DefaultQuery<I, Word<O>> query = queries.get(i);
+                    final int oracleIndex = i % sulOracles.size();
```
fixed

c-southwest commented on src/main/java/com/github/protocolfuzzing/protocolstatefuzzer/components/learner/oracles/RandomWpMethodEQOracle.java:
```diff
@@ -127,28 +135,61 @@ public RandomWpMethodEQOracle(MealyMembershipOracle<I, O> sulOracle,
+                                return query;  // Find counterexample
+                            }
+                        } catch (Exception e) {
+                            System.err.println("[ERROR] process query: " + e.getMessage());
```
fixed

c-southwest commented:
I fixed all the comments, and I pushed a new DTLS-Fuzzer commit that uses the latest parallel PSF:
https://github.com/assist-project/dtls-fuzzer/actions/runs/15587867792

c-southwest commented:
> > ... I pushed a new DTLS-Fuzzer commit that uses the latest parallel PSF: https://github.com/c-southwest/dtls-fuzzer/actions/runs/15587976748
> 
> Something did not work OK in that CI run. Was the new command line option renamed? (`-threadCount` vs `-equivalenceThreadCount`, which is a very long name by the way and better be shortened).

Fixed. I added a shorter alternetive parameter "-eqvThreads".
CI Tests of DTLS-Fuzzer with latest PSF commit: https://github.com/assist-project/dtls-fuzzer/actions/runs/15592121847/job/43913533056

actyp commented:
That sounds optimal to me. Here are my thoughts.

Let's say that we run our equivalence queries in parallel and we find more than one counterexamples to the current hypothesis.

According to our deterministic constraint it suffices to keep the counterexample that the thread with the smallest ticket number found. This holds in our case, because we have been using a default `seed` in our `LearnerConfigStandard`, which is fed into the `Random` generator. This also means that the whole sequence of equivalence queries and their results will be deterministic with respect to the same seed.  

#### Comment 1

It is possible if we change the seed, that the equivalence phase would be faster (or slower).

#### Comment 2

I think we can utilize the remaining counterexamples by calling the `learner.refineHypothesis(counterExample)` in `StateFuzzerStandard` multiple times. The [learlib api](https://learnlib.de/learnlib/maven-site/0.18.0/apidocs/de.learnlib.api/de/learnlib/algorithm/LearningAlgorithm.html#refineHypothesis(de.learnlib.query.DefaultQuery)) says that this method returns a boolean indicating whether the provided counterexample refined the model or not, which means that we can also call it with a **possible** counterexample instead of a **certain** one.

The question in that case is whether the multiple refinements would have a net benefit or not. We can leave that experimentation for a different PR if we want.

#### Comment 3

From our argument files in DTLS-Fuzzer and EDHOC-Fuzzer we usually use more than one equivalence algorithms, which I assume run sequentially upon failure.

Also, I just noticed that in DTLS-Fuzzer we use the `RANDOM_WP_METHOD` as an equivalence algorithm but not in EDHOC-Fuzzer.

kostis commented:
> @kostis I implemented your proposal, and the resulting code base is much simpler, and this new parallel scheme can be used with many threads without extra problems (For example, using 100 tasks at a time is not effictive if we are using 128 or 256 threads).
> 
> Additionally, now we don't need to generate queries beforehand, we can generate a new query each time the thread gets a new ticket number, so we can remove List or Array to save some memory.
> 
> Lastly, with this new design, Thread Pool becomes unnecessary, so I remove it and use the traditional way instead.

Great news and great work @c-southwest !

Nice that this turned out easy to implement, cleaner, and also overall a simpler scheme.  Unless I hear some objection during the weekend, I will merge this on Monday.

If I am not mistaken, there is no fundamental reason why the same parallelization scheme cannot be used for all other equivalence algorithms (e.g. `WP_METHOD` , `W_METHOD`, etc.).  In the spirit of not slowing down development, I suggest we leave this for another PR, but if @c-southwest or anybody else has some cycles to spare, I suggest they go ahead and implement this.

Let me also quickly reply to the three comments of @actyp :
1. Yes, of course.
2. Interesting thought, and we should implement this, but let's leave this for another PR (as you also suggest).
3. Our DTLS-Fuzzer experience is that `RANDOM_WP_METHOD` is usually more effective in discovering counterexamples in SMs with many states, so perhaps we should investigate if it makes any significant difference for EDHOC-Fuzzer too, esp. since that algorithm has now been parallelized.
